include/maths/CBoostedTreeImpl.h

```diff
@@ -0,0 +1,698 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#ifndef INCLUDED_ml_maths_CBoostedTreeImpl_h
+#define INCLUDED_ml_maths_CBoostedTreeImpl_h
+
+#include <core/CContainerPrinter.h>
+#include <core/CDataFrame.h>
+#include <core/CPackedBitVector.h>
+
+#include <maths/CBasicStatistics.h>
+#include <maths/CBayesianOptimisation.h>
+#include <maths/CBoostedTree.h>
+#include <maths/CDataFrameUtils.h>
+#include <maths/CLinearAlgebraEigen.h>
+#include <maths/CQuantileSketch.h>
+#include <maths/CSampling.h>
+#include <maths/CTools.h>
+#include <maths/ImportExport.h>
+
+#include <boost/operators.hpp>
+#include <boost/optional.hpp>
+
+#include "COutliers.h"
+#include <numeric>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+namespace ml {
+namespace maths {
+
+namespace {
+const double INF{std::numeric_limits<double>::max()};
+}
+
+class CBoostedTreeImpl final {
+
+public:
+    using TMeanAccumulator = CBasicStatistics::SSampleMean<double>::TAccumulator;
+    using TMeanVarAccumulator = CBasicStatistics::SSampleMeanVar<double>::TAccumulator;
+    using TBayesinOptimizationUPtr = std::unique_ptr<maths::CBayesianOptimisation>;
+    using TDoubleVec = std::vector<double>;
+
+public:
+    static const double MINIMUM_ETA;
+    static const std::size_t MAXIMUM_NUMBER_TREES;
+    static const double MINIMUM_RELATIVE_GAIN_PER_SPLIT;
+
+public:
+    CBoostedTreeImpl(std::size_t numberThreads,
+                     std::size_t dependentVariable,
+                     CBoostedTree::TLossFunctionUPtr loss);
+    //! Set the number of folds to use for estimating the generalisation error.
+    void numberFolds(std::size_t numberFolds);
+    //! Set the lambda regularisation parameter.
+    void lambda(double lambda);
+    //! Set the gamma regularisation parameter.
+    void gamma(double gamma);
+    //! Set the amount we'll shrink the weights on each each iteration.
+    void eta(double eta);
+    //! Set the maximum number of trees in the ensemble.
+    void maximumNumberTrees(std::size_t maximumNumberTrees);
+    //! Set the fraction of features we'll use in the bag to build a tree.
+    void featureBagFraction(double featureBagFraction);
+    //! Set the maximum number of optimisation rounds we'll use for hyperparameter
+    //! optimisation per parameter.
+    void maximumOptimisationRoundsPerHyperparameter(std::size_t rounds);
+
+    //! Train the model on the values in \p frame.
+    void train(core::CDataFrame& frame, CBoostedTree::TProgressCallback recordProgress);
```

Comment:
```suggestion
    void train(core::CDataFrame& frame, CBoostedTree::TProgressCallback recordProgress);
    
```
Also, given you pass frame to the factory it seems slightly odd that you pass it again to these functions. In particular, this has to be the same frame that you built it on so this seems like a recipe for abuse.

Reply:
I'm thinking we should move towards storing a reference to the frame to run on in the base class, having concrete `void train(CBoostedTree::TProgressCallback recordProgress)` and a private `virtual void trainImpl(core::CDataFrame& frame, CBoostedTree::TProgressCallback recordProgress)` which can then be overridden in the derived classes. And similarly for other interface functions which currently take a `frame`.

Reply:
done see PR #555